### PR TITLE
Martin's Comments

### PR DESCRIPTION
**Pull request by mrdeep1**

### CODE REVIEW

File: draft-ietf-core-new-block.xml [hunk 1]
```diff
@@ -1002,9 +1003,15 @@ missing-block-number = uint
...
-        <t>NON_TIMEOUT is the period of delay between sending MAX_PAYLOADS_SET
-        for the same body. By default, NON_TIMEOUT has the same value as
-        ACK_TIMEOUT (Section 4.8 of <xref target="RFC7252"></xref>).</t>
+        <t>NON_TIMEOUT is used to calculate the delay between sending
+        MAX_PAYLOADS_SET for the same body. By default, NON_TIMEOUT has the
+        same value as ACK_TIMEOUT (Section 4.8 of <xref
+        target="RFC7252"></xref>).</t>
+
+        <t>NON_TIMEOUT_RANDOM is the actual delay between sending
+        MAX_PAYLOADS_SET of the same body which is of a random duration (often
```
Comment (boucadair) [inline draft-ietf-core-new-block.xml]:
```suggestion
        MAX_PAYLOADS_SET of the same body. It is a random duration (
```

File: draft-ietf-core-new-block.xml [hunk 2]
```diff
@@ -1002,9 +1003,15 @@ missing-block-number = uint
...
+        not an integral number of seconds) between NON_TIMEOUT and (NON_TIMEOUT
+        * ACK_RANDOM_FACTOR).</t>
```
Comment (boucadair) [inline draft-ietf-core-new-block.xml]:
```suggestion
        * ACK_RANDOM_FACTOR). ACK_RANDOM_FACTOR is set to 1.5 as discussed in Section 4.8 of <xref target="RFC7252"></xref>.</t>
```

File: draft-ietf-core-new-block.xml [hunk 3]
```diff
@@ -1002,9 +1003,15 @@ missing-block-number = uint
             target="scope"></xref>.</t>
           </list></t>
 
-        <t>NON_TIMEOUT is the period of delay between sending MAX_PAYLOADS_SET
-        for the same body. By default, NON_TIMEOUT has the same value as
-        ACK_TIMEOUT (Section 4.8 of <xref target="RFC7252"></xref>).</t>
+        <t>NON_TIMEOUT is used to calculate the delay between sending
```
Comment (boucadair) [inline draft-ietf-core-new-block.xml]:
```suggestion
        <t>NON_TIMEOUT is used to compute the delay between sending
```

File: draft-ietf-core-new-block.xml [hunk 4]
```diff
@@ -1029,20 +1036,25 @@ missing-block-number = uint
...
 
         <t>NON_PROBING_WAIT is used to limit the potential wait needed when
-        using PROBING_RATE. By default, NON_PROBING_WAIT is computed in the
-        same way as EXCHANGE_LIFETIME (Section 4.8.2 of <xref
-        target="RFC7252"></xref>) but with ACK_TIMEOUT and MAX_RETRANSMIT
-        substituted with NON_TIMEOUT and NON_MAX_RETRANSMIT,
-        respectively:<list style="empty">
+        using PROBING_RATE. By default, NON_PROBING_WAIT is computed in a
+        similar way as EXCHANGE_LIFETIME (Section 4.8.2 of <xref
+        target="RFC7252"></xref>) but with ACK_TIMEOUT, MAX_RETRANSMIT and
```
Comment (boucadair) [inline draft-ietf-core-new-block.xml]:
```suggestion
        target="RFC7252"></xref>) but with ACK_TIMEOUT, MAX_RETRANSMIT, and
```

File: draft-ietf-core-new-block.xml [hunk 5]
```diff
@@ -1029,20 +1036,25 @@ missing-block-number = uint
...
+        PROCESSING_DELAY substituted with NON_TIMEOUT, NON_MAX_RETRANSMIT and
```
Comment (boucadair) [inline draft-ietf-core-new-block.xml]:
```suggestion
        PROCESSING_DELAY substituted with NON_TIMEOUT, NON_MAX_RETRANSMIT, and
```

File: draft-ietf-core-new-block.xml [hunk 6]
```diff
@@ -1029,20 +1036,25 @@ missing-block-number = uint
...
+        NON_TIMEOUT_RANDOM, respectively of the EXCHANGE_LIFETIME definition
```
Comment (boucadair) [inline draft-ietf-core-new-block.xml]:
```suggestion
        NON_TIMEOUT_RANDOM, respectively:
```

File: draft-ietf-core-new-block.xml [hunk 7]
```diff
@@ -1029,20 +1036,25 @@ missing-block-number = uint
...
+        expansion:<list style="empty">
```
Comment (boucadair) [inline draft-ietf-core-new-block.xml]:
```suggestion
        <list style="empty">
```

File: draft-ietf-core-new-block.xml [hunk 8]
```diff
@@ -1029,20 +1036,25 @@ missing-block-number = uint
...
           </list></t>
 
         <t>NON_PARTIAL_TIMEOUT is used for expiring partially received bodies.
         By default, NON_PARTIAL_TIMEOUT is computed in the same way as
-        EXCHANGE_LIFETIME (Section 4.8.2 of <xref target="RFC7252"></xref>).
-        This default value is calculated in the same way as
-        NON_PROBING_WAIT.</t>
+        EXCHANGE_LIFETIME (Section 4.8.2 of <xref target="RFC7252"></xref>) but
+        with ACK_TIMEOUT and MAX_RETRANSMIT substituted with NON_TIMEOUT and
+        NON_MAX_RETRANSMIT, respectively of the EXCHANGE_LIFETIME definition
```
Comment (boucadair) [inline draft-ietf-core-new-block.xml]:
```suggestion
        NON_MAX_RETRANSMIT, respectively
```

File: draft-ietf-core-new-block.xml [hunk 9]
```diff
@@ -1029,20 +1036,25 @@ missing-block-number = uint
...
+        expansion:<list style="empty">
```
Comment (boucadair) [inline draft-ietf-core-new-block.xml]:
```suggestion
        :<list style="empty">
```

File: draft-ietf-core-new-block.xml [hunk 10]
```diff
@@ -1029,20 +1036,25 @@ missing-block-number = uint
...
+            <t>NON_PROBING_WAIT = NON_TIMEOUT * ((2 ** NON_MAX_RETRANSMIT) -
```
Comment (boucadair) [inline draft-ietf-core-new-block.xml]:
```suggestion
            <t>NON_PARTIAL_TIMEOUT = NON_TIMEOUT * ((2 ** NON_MAX_RETRANSMIT) -
```

File: draft-ietf-core-new-block.xml [hunk 11]
```diff
@@ -1070,12 +1083,9 @@ Table 3: Congestion Control Parameters]]></artwork>
             other transmission parameters are negotiated between peers. Even
             when not negotiated, the DOTS application uses customized defaults
             as discussed in Section 4.5.2 of <xref target="RFC8782"></xref>.
-            Note that MAX_PAYLOADS, NON_MAX_RETRANSMIT, NON_TIMEOUT,
-            NON_PROBING_WAIT, and NON_PARTIAL_TIMEOUT can be negotiated
-            between DOTS peers, e.g., as per <xref
-            target="I-D.bosh-dots-quick-blocks"></xref>. When explicit values
-            are configured for NON_PROBING_WAIT and NON_PARTIAL_TIMEOUT, these
-            values are used without applying any jitter. </t>
+            Note that MAX_PAYLOADS, NON_MAX_RETRANSMIT, and NON_TIMEOUT can be
```
Comment (boucadair) [inline draft-ietf-core-new-block.xml]: Please revert to the OLD one.

File: draft-ietf-core-new-block.xml [hunk 12]
```diff
@@ -1002,9 +1003,15 @@ missing-block-number = uint
...
-        for the same body. By default, NON_TIMEOUT has the same value as
-        ACK_TIMEOUT (Section 4.8 of <xref target="RFC7252"></xref>).</t>
+        <t>NON_TIMEOUT is used to calculate the delay between sending
+        MAX_PAYLOADS_SET for the same body. By default, NON_TIMEOUT has the
+        same value as ACK_TIMEOUT (Section 4.8 of <xref
+        target="RFC7252"></xref>).</t>
+
+        <t>NON_TIMEOUT_RANDOM is the actual delay between sending
+        MAX_PAYLOADS_SET of the same body which is of a random duration (often
+        not an integral number of seconds) between NON_TIMEOUT and (NON_TIMEOUT
```
Comment (boucadair) [inline draft-ietf-core-new-block.xml]: We need to mention that once a random value is computed, that value does not change for a block-wise transfer.

File: draft-ietf-core-new-block.xml [hunk 13]
```diff
@@ -986,8 +986,9 @@ missing-block-number = uint
 
       <section anchor="cc-non" title="Non-confirmable (NON)">
         <t>This document introduces new parameters MAX_PAYLOADS, NON_TIMEOUT,
-        NON_RECEIVE_TIMEOUT, NON_MAX_RETRANSMIT, NON_PROBING_WAIT, and
-        NON_PARTIAL_TIMEOUT primarily for use with NON (Table 3).</t>
+        NON_TIMEOUT_RANDOM, NON_RECEIVE_TIMEOUT, NON_MAX_RETRANSMIT,
```
Comment (boucadair) [inline draft-ietf-core-new-block.xml:989]: I wonder whether we need to add a note to explain that randomness may be naturally provided with the traffic profile, how probing-rate is computed (as this is an avg) and the specifics in 3.2. Adding randomness cover cases where all is in sync.

File: draft-ietf-core-new-block.xml [hunk 14]
```diff
@@ -1051,8 +1063,9 @@ missing-block-number = uint
 | MAX_PAYLOADS        |            10 |
 | NON_MAX_RETRANSMIT  |             4 |
 | NON_TIMEOUT         |           2 s |
+| NON_TIMEOUT_RANDOM  |         2-3 s |
```
Comment (boucadair) [inline draft-ietf-core-new-block.xml]:
```suggestion
| NON_TIMEOUT_RANDOM  |         between 2-3 s |
```

File: draft-ietf-core-new-block.xml [hunk 15]
```diff
@@ -1051,8 +1063,9 @@ missing-block-number = uint
...
 | NON_RECEIVE_TIMEOUT |           4 s |
-| NON_PROBING_WAIT    |         247 s |
+| NON_PROBING_WAIT    |     247-248 s |
```
Comment (boucadair) [inline draft-ietf-core-new-block.xml]:
```suggestion
| NON_PROBING_WAIT    |    between 247-248 s |
```

**Review by boucadair — CHANGES_REQUESTED** [review]

File: draft-ietf-core-new-block.xml (hunk 1 above)
Comment (mrdeep1) [reply draft-ietf-core-new-block.xml]: Updated

File: draft-ietf-core-new-block.xml (hunk 2 above)
Comment (mrdeep1) [reply draft-ietf-core-new-block.xml]: Updated

File: draft-ietf-core-new-block.xml (hunk 3 above)
Comment (mrdeep1) [reply draft-ietf-core-new-block.xml]: Updated

File: draft-ietf-core-new-block.xml (hunk 4 above)
Comment (mrdeep1) [reply draft-ietf-core-new-block.xml]: Updated

File: draft-ietf-core-new-block.xml (hunk 5 above)
Comment (mrdeep1) [reply draft-ietf-core-new-block.xml]: Updated

File: draft-ietf-core-new-block.xml (hunk 7 above)
Comment (mrdeep1) [reply draft-ietf-core-new-block.xml]: Updated

File: draft-ietf-core-new-block.xml (hunk 8 above)
Comment (mrdeep1) [reply draft-ietf-core-new-block.xml]: Updated

File: draft-ietf-core-new-block.xml (hunk 9 above)
Comment (mrdeep1) [reply draft-ietf-core-new-block.xml]: Updated

File: draft-ietf-core-new-block.xml (hunk 10 above)
Comment (mrdeep1) [reply draft-ietf-core-new-block.xml]: Oops - corrected.

File: draft-ietf-core-new-block.xml (hunk 11 above)
Comment (mrdeep1) [reply draft-ietf-core-new-block.xml]: Reverted

File: draft-ietf-core-new-block.xml (hunk 12 above)
Comment (mrdeep1) [reply draft-ietf-core-new-block.xml]: Actually, it needs to be random per MAX_PAYLOADS_SET.  E.g. MAX_PAYLOADS_SET#1 from 2 transmitters are slightly out of sync.  MAX_PAYLOADS_SET#2 is in sync.  Need to make sure that MAX_PAYLOADS_SET#3 does not remain in sync. Made this more explicit as
````
-    <t>NON_TIMEOUT_RANDOM is the actual delay between sending
+    <t>NON_TIMEOUT_RANDOM is the actual delay between sending each
````

Comment (boucadair) [reply draft-ietf-core-new-block.xml]: If we have it random in the first set, that randomization will apply for all sets. What would be the benefit to re-randomize for each set?

Comment (mrdeep1) [reply draft-ietf-core-new-block.xml]: Fair point - no need for wasting random compute cycles.  My updated implementation just grabbed a new random value for each MAX_PAYLOADS_SET, but I guess there is no reason why the first generated random value cannot be saved and used subsequently for the same body.
````
-        <t>NON_TIMEOUT_RANDOM is the actual delay between sending
-        MAX_PAYLOADS_SET of the same body which is of a random duration (often
+        <t>NON_TIMEOUT_RANDOM is the initial actual delay between sending the
+        first two MAX_PAYLOADS_SETs of the same body. The same delay is then
+        used between the subsequent MAX_PAYLOADS_SETs. It is a random duration (
````

File: draft-ietf-core-new-block.xml (hunk 14 above)
Comment (mrdeep1) [reply draft-ietf-core-new-block.xml]: Updated

File: draft-ietf-core-new-block.xml (hunk 15 above)
Comment (mrdeep1) [reply draft-ietf-core-new-block.xml]: Updated

File: draft-ietf-core-new-block.xml (hunk 12 above)
Comment (boucadair) [reply draft-ietf-core-new-block.xml]: Thanks. Let's make the intent clear in the text (randomize in the first set and then reuse that value for subsequent sets of the same transfer)

**Comment by mrdeep1** [comment]
Changes pushed

**Comment by mrdeep1** [comment]
Fixed a couple of formatting errors - code re-pushed.

File: draft-ietf-core-new-block.xml [hunk 16]
```diff
@@ -986,8 +986,15 @@ missing-block-number = uint
 
       <section anchor="cc-non" title="Non-confirmable (NON)">
         <t>This document introduces new parameters MAX_PAYLOADS, NON_TIMEOUT,
-        NON_RECEIVE_TIMEOUT, NON_MAX_RETRANSMIT, NON_PROBING_WAIT, and
-        NON_PARTIAL_TIMEOUT primarily for use with NON (Table 3).</t>
+        NON_TIMEOUT_RANDOM, NON_RECEIVE_TIMEOUT, NON_MAX_RETRANSMIT,
+        NON_PROBING_WAIT, and NON_PARTIAL_TIMEOUT primarily for use with NON
+        (Table 3).<list style="empty">
+            <t>Note: Randomness may naturally be provided based on the traffic
+            profile, how probing-rate is computed (as this is an average) and
```
Comment (boucadair) [inline draft-ietf-core-new-block.xml]:
```suggestion
            profile, how probing-rate is computed (as this is an average), and
```

File: draft-ietf-core-new-block.txt [hunk 17]
```diff
@@ -1104,31 +1104,45 @@ Internet-Draft      Quick Block-Wise Transfer Options           May 2021
...
+   (Table 3).
 
-   MAX_PAYLOADS should be configurable with a default value of 10.  Both
-   CoAP endpoints MUST have the same value (otherwise there will be
-   transmission delays in one direction) and the value MAY be negotiated
-   between the endpoints to a common value by using a higher level
-   protocol (out of scope of this document).  This is the maximum number
-   of payloads that can be transmitted at any one time.
+      Note: Randomness may naturally be provided based on the traffic
+      profile, how probing-rate is computed (as this is an average) and
```
Comment (boucadair) [inline draft-ietf-core-new-block.txt]:
```suggestion
      profile, how probing-rate is computed (as this is an average), and
```

File: draft-ietf-core-new-block.xml (hunk 13 above)
Comment (mrdeep1) [reply draft-ietf-core-new-block.xml:989]: Added as a starting point.
````
<list style="empty">
            <t>Note: Randomness may naturally be provided based on the traffic 
            profile, how probing-rate is computed (as this is an average) and
            when the peer responds. Randomness is explicitly added for some of
            the congestion control parameters to handle situations where every 
            thing is in sync when retrying.</t>
          </list>
````

**Pull request merged**